DP-38621: Enable microsite-scoped searching

andy-blum opened the pull request:
**Description:**
This PR allows pages within a microsite to use search.mass.gov while scoping results within the microsite.


**Jira:** (Skip unless you are MA staff)
DP-38621

**To Test:**
- [ ] Add steps to test this feature


---

[Peer Review Checklist](https://github.com/massgov/openmass/blob/develop/docs/peer_review_checklist.md)

andy-blum committed:
WIP: modify searchHeader parameters within microsites

andy-blum committed:
DP-38621: update search form scoping

andy-blum committed:
DP-38621: add changelog

andy-blum committed:
DP-38621: fix phpcs error

andy-blum committed:
DP-38621: remove search.js from mass_microsites

dstorozhuk commented:
@andy-blum  we have time to update the tests in the scope of this PR, so please do so.

andy-blum committed:
DP-38621: add test coverage

andy-blum committed:
DP-38621: update tests to submit on form, not input

andy-blum commented:
@dstorozhuk test coverage is added and passing.

andy-blum committed:
DP-38621: update tests & mayflower deps

andy-blum committed:
DP-38621: show microsite scope on first suggestion

andy-blum committed:
DP-38621: add 'in' to org labels

andy-blum committed:
DP-38621: prevent duplicate scopes

dstorozhuk commented on docroot/modules/custom/mayflower/src/Prepare/Molecules.php:
```diff
@@ -1269,34 +1269,66 @@ public static function prepareGoogleMapSection($entity) {
+    $show_default_scope = TRUE;
+    $default_scope = [];
     $suggested_scopes = [];
+    $placeholder = 'Search Mass.gov';
```
@andy-blum  please use string translation trait.

dstorozhuk commented on docroot/modules/custom/mayflower/src/Prepare/Molecules.php:
```diff
@@ -1269,34 +1269,66 @@ public static function prepareGoogleMapSection($entity) {
+        ];
+        $placeholder = 'Search ' . $microsite->label();
+        $suggested_scopes[] = [
+          'label' => 'in all of Mass.gov',
```
@andy-blum  please use string translation trait.

dstorozhuk commented on docroot/modules/custom/mayflower/src/Prepare/Molecules.php:
```diff
@@ -1269,34 +1269,66 @@ public static function prepareGoogleMapSection($entity) {
+        $default_scope = [
+          'value' => str_replace("-", "", $utilities->slugify(trim($microsite->label()))),
+          'type' => 'microsite',
+          'label' => 'in ' . $microsite->label(),
```
@andy-blum  please use string translation trait.

dstorozhuk commented on docroot/modules/custom/mayflower/src/Prepare/Molecules.php:
```diff
@@ -1269,34 +1269,66 @@ public static function prepareGoogleMapSection($entity) {
+                    $cache_tags = array_merge($cache_tags, $parent->getCacheTags());
+                    $parent_slug = $utilities->slugify(trim($parent->label()));
+                    $suggested_scopes[$parent_slug] = [
+                      'label' => "in " . trim($parent->label()),
```
@andy-blum  please use string translation trait.

dstorozhuk commented on docroot/modules/custom/mayflower/src/Prepare/Molecules.php:
```diff
@@ -1269,34 +1269,66 @@ public static function prepareGoogleMapSection($entity) {
+                $cache_tags = array_merge($cache_tags, $org->getCacheTags());
+                $slug = $utilities->slugify(trim($org->label()));
+                $suggested_scopes[$slug] = [
+                  'label' => "in " . trim($org->label()),
```
@andy-blum  please use string translation trait.

dstorozhuk commented on docroot/modules/custom/mayflower/src/Prepare/Molecules.php:
```diff
@@ -1306,15 +1338,17 @@ public static function prepareHeaderSearch(?object $entity = NULL, array &$cache
       'name' => 'header-search',
       'id' => 'header-search',
-      'placeholder' => 'Search Mass.gov',
+      'placeholder' => $placeholder,
```
@andy-blum  please make sure the strings are wrapped in one of the existing translation methods. In the previous version it was easy for use to wrap the placeholder to as string. Now it is not. We potentially can enable interface translation in a future.

andy-blum committed:
DP-38621: add string transtlation  trait

dstorozhuk commented on docroot/modules/custom/mayflower/src/Prepare/Molecules.php:
```diff
@@ -1272,7 +1272,7 @@ public static function prepareHeaderSearch(?object $entity = NULL, array &$cache
     $default_scope = [];
     $suggested_scopes = [];
-    $placeholder = 'Search Mass.gov';
+    $placeholder = t('Search Mass.gov');
```
@andy-blum  I mean this https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21StringTranslation%21StringTranslationTrait.php/trait/StringTranslationTrait/10

andy-blum replied:
@dstorozhuk we cannot use StringTranslationTrait here, as all methods in this class are static methods. Looking around, I can see `\Drupal::translaton()->translate();` but that method has this warning:

```php
  /**
   * Never call this translate() method directly. In order for strings to be
   * localized, make them available in one of the ways supported by the
   * @link https://www.drupal.org/node/322729 Localization API @endlink. When
   * possible, use the \Drupal\Core\StringTranslation\StringTranslationTrait
   * $this->t(). Otherwise create a new
   * \Drupal\Core\StringTranslation\TranslatableMarkup object.
   */
```

We could replace usages of `t()` with new TranslatableMarkup objects, but that's just avoiding the use of a global function to do identical work at that point:

```php
function t($string, array $args = [], array $options = []) {
  return new TranslatableMarkup($string, $args, $options);
}
```

dstorozhuk committed:
DP-38621: Update Mayflower artifacts dependency to the develop branch

Switched the `massgov/mayflower-artifacts` dependency from a feature-specific branch to the `dev-develop` branch in `composer.json`. Updated the corresponding reference and metadata in `composer.lock` to reflect the latest changes.

Signed-off-by: Dmytro Storozhuk <dima@itech4web.com>
